extensions/ql-vscode/src/view/results/AlertTablePathNodeRow.tsx

```diff
@@ -9,6 +17,151 @@ import { SarifLocation } from "./locations/SarifLocation";
 import { selectableZebraStripe } from "./result-table-utils";
 import { useCallback, useMemo } from "react";
 import { VerticalRule } from "../common/VerticalRule";
+import type { UserSettings } from "../../common/interface-types";
+
+/** The definition of a taxon for a data extension model row. */
+interface ModelTaxon {
+  location: SarifLogLocation;
+}
+
+/** Resolve an `ArtifactLocation` that might contain a relative reference instead of an absolute
+ * URI.
+ */
+function resolveArtifactLocation(
+  location: ArtifactLocation,
+  baseUri: URL,
+): ArtifactLocation {
+  if (location.uri === undefined) {
+    // No URI at all. Just return the original location.
+    return location;
+  }
+  return {
+    ...location,
+    uri: new URL(location.uri, baseUri).toString(),
+  };
+}
+
+/** Get the URI of the pack's local root directory, if available. */
+function getLocalPackUri(extension: ToolComponent): URL | undefined {
+  if (extension.locations === undefined) {
+    return undefined;
+  }
+
+  const localPackLocation = extension.locations.find(
+    (loc) =>
+      loc.properties !== undefined &&
+      loc.properties.tags !== undefined &&
+      loc.properties.tags.includes("CodeQL/LocalPackRoot"),
+  );
+  if (localPackLocation === undefined || localPackLocation.uri === undefined) {
+    return undefined;
+  }
+
+  return new URL(localPackLocation.uri);
+}
+
+/** Resolve a `ReportingDescriptorReference` to the `ReportingDescriptor` for the taxon that it
+ * refers to.
+ */
+function resolveTaxonDefinition(
+  run: Run,
+  taxonRef: ReportingDescriptorReference,
+): ModelTaxon | undefined {
+  const extensions = run.tool.extensions;
+  if (extensions === undefined) {
+    return undefined;
+  }
+
+  const extensionIndex = taxonRef.toolComponent?.index;
+  if (
+    extensionIndex === undefined ||
+    extensionIndex < 0 ||
+    extensionIndex >= extensions.length
+  ) {
+    return undefined;
+  }
+
+  const extension = extensions[extensionIndex];
+  if (extension.taxa === undefined) {
+    return undefined;
+  }
+
+  const localPackUri = getLocalPackUri(extension);
+  if (localPackUri === undefined) {
+    return undefined;
+  }
+
+  const taxonIndex = taxonRef.index;
+  if (
+    taxonIndex === undefined ||
+    taxonIndex < 0 ||
+    taxonIndex >= extension.taxa.length
+  ) {
+    return undefined;
+  }
+
+  const taxonDef = extension.taxa[taxonIndex];
+  if (taxonDef.properties === undefined) {
+    return undefined;
+  }
+
+  const location: PhysicalLocation =
+    taxonDef.properties["CodeQL/DataExtensionLocation"];
+  if (location === undefined || location.artifactLocation === undefined) {
+    return undefined;
+  }
+
+  return {
+    location: {
+      physicalLocation: {
+        ...location,
+        artifactLocation: resolveArtifactLocation(
+          location.artifactLocation,
+          localPackUri,
+        ),
+      },
+    },
+  };
+}
+
+/** Generate the React elements for each taxon. */
+function taxaLocations(
```

Comment:
Since this returns a React node, it would probably be better to actually make this a React component so it can be used like `<TaxaLocations taxa={step.taxa} run={run} onClick={handleSarifLocationClicked} />`. You can probably also move this to a separate file to make this file smaller and more contained.

Reply:
Good call. Done.